src/common/ipc.ts

```diff
@@ -2,94 +2,133 @@
 import logger from "../main/logger";
+import { getRandId } from "./utils";
 
 export type IpcChannel = string;
 
-export interface IpcHandleOpts {
-  timeout?: number;
+export enum IpcMode {
+  SYNC = "sync",
+  ASYNC = "async",
 }
 
-export interface IpcMessageHandler<T extends any[] = any> {
-  (...args: T): any;
+export interface IpcChannelRequest<A extends any[] = any> {
+  msgId: string;
+  args: A;
 }
 
-export interface IpcMessageOpts<A extends any[] = any> {
+export interface IpcChannelResponse<T extends any[] = any, E = any> {
+  msgId: string;
+  data?: T;
+  error?: E;
+}
+
+export interface IpcChannelInit {
+  channel: IpcChannel; // main <-> renderer communication channel name
+  mode?: IpcMode; // default: "async", use "sync" as last resort: https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args
+  handle?: (...args: any[]) => any; // main-process message handler
+  autoBind?: boolean; // auto-bind message handler in main-process, default: false
+  timeout?: number; // timeout for waiting response from the sender
+  once?: boolean; // todo: add support
+}
+
+export function createIpcChannel({ autoBind = true, mode = IpcMode.ASYNC, timeout = 0, handle, channel }: IpcChannelInit) {
+  channel = `${mode}:${channel}`
+
+  const ipcChannel = {
+    channel: channel,
+    handleInMain: () => {
+      logger.info(`[IPC]: setup channel "${channel}"`);
+
+      ipcMain.on(channel, async (event, req: IpcChannelRequest) => {
+        let resolved = false;
+        let timerId: any;
+
+        function resolve(res: Partial<IpcChannelResponse>) {
+          if (resolved) return;
+          res.msgId = req.msgId; // return back to sender to be able to handle response
+          resolved = true
+          logger.debug(`[IPC]: sending response to "${channel}"`, res);
+          if (mode === IpcMode.ASYNC) {
+            event.reply(channel, res);
+          }
+          if (mode === IpcMode.SYNC) {
+            event.returnValue = res;
+          }
+        }
+
+        if (timeout > 0) {
+          timerId = setTimeout(() => {
+            const timeoutError = new Error(`[IPC]: response timeout in ${timeout}ms`);
+            resolve({ error: timeoutError })
+          }, timeout);
+        }
+
+        try {
+          const data = await handle(...req.args); // todo: maybe exec in separate thread/worker
+          resolve({ data })
+        } catch (error) {
+          resolve({
+            error: String(error)
+          })
+        } finally {
+          clearTimeout(timerId);
+        }
+      })
+    },
+    invokeFromRenderer: async (...args: any[]) => {
+      const req: IpcChannelRequest = {
+        msgId: getRandId({ prefix: "ipc-msg-id" }),
+        args: args,
+      }
+      if (mode === IpcMode.ASYNC) {
+        ipcRenderer.send(channel, req)
+      }
+      if (mode === IpcMode.SYNC) {
+        ipcRenderer.sendSync(channel, req)
+      }
+      return new Promise(async (resolve, reject) => {
```

Comment:
why are we returning a promise if the mode is `sync`?

Reply:
maybe we should, haven't even tested `sync` mode and it's better to use this "as last resort" as defined in docs:
https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args

Reply:
Just thinking that do we need `sync` mode here if we are only using async messaging?

Reply:
Well, this `IPC` helper is intended to be multi-functional, so I would keep it, but have to test if it's really works as expected.